File: facilitator-solutions/csharp/rung-5-bitset.cs
using System;
using System.Collections.Generic;

namespace Race;

// Rung 5 - bit packing. Includes rungs 1 to 4.
//
// One bit per odd candidate, packed into 64-bit words. Eight times smaller again,
// sixteen times smaller than the original byte-per-number array.
//
//     word index    i >> 6
//     bit within    i & 63
//     test          (words[i >> 6] & (1UL << i)) != 0
//     clear         words[i >> 6] &= ~(1UL << i)
//
// C# masks the shift count for you, so `1UL << i` behaves as `1UL << (i & 63)`.
// Go does not do this, which is why the Go version has to write the mask out.
//
// This is NOT faster per operation - a bit write is a read, a shift, an or and a
// write where a byte write is a single store. It wins because it does half the
// operations in a sixteenth of the space.
//
// Here a set bit still means prime, so the words are filled with ones. The tail
// word holds bits beyond _count; they are filled too, and Result is bounded by
// _count so they are never read. Rung 6 inverts this, and that is where the tail
// word turns into a bug.
public class Sieve
{
	private const int WordBits = 64;

	private readonly int _sieveSize;
	private readonly int _count;
	private readonly ulong[] _words;

	public Sieve(int sieveSize)
	{
		_sieveSize = sieveSize;
		_count = (sieveSize + 1) / 2;
		_words = new ulong[(_count + WordBits - 1) / WordBits];
		Array.Fill(_words, ulong.MaxValue);
	}

	public void Run()
	{
		int limit = (int)Math.Sqrt(_sieveSize);

		for (int factor = 3; factor <= limit; factor += 2)
		{
			int f = factor >> 1;

			if ((_words[f >> 6] & (1UL << f)) != 0)
			{
				for (int q = (factor * factor) >> 1; q < _count; q += factor)
				{
					_words[q >> 6] &= ~(1UL << q);
				}
			}
		}
	}

	public List<int> Result()
	{
		List<int> primes = [];

		if (_sieveSize >= 2)
		{
			primes.Add(2);
		}

		for (int i = 1; i < _count; i++)
		{
			if ((_words[i >> 6] & (1UL << i)) != 0)
			{
				primes.Add(2 * i + 1);
			}
		}

		return primes;
	}
}

File: facilitator-solutions/fsharp/rung-5-bitset.fs
namespace DragRace

// Rung 5 - bit packing. Includes rungs 1 to 4.
//
// One bit per odd candidate in 64-bit words. Eight times smaller again, sixteen
// times smaller than the original.
//
//     word index    i >>> 6
//     bit within    i &&& 63
//     test          words.[i >>> 6] &&& (1UL <<< i) <> 0UL
//     clear         words.[i >>> 6] <- words.[i >>> 6] &&& ~~~(1UL <<< i)
//
// .NET masks the shift count, so 1UL <<< i behaves as 1UL <<< (i &&& 63).
//
// A set bit still means prime here, so the words are filled with ones and the tail
// word's spare bits are set too. getResults is bounded by count, so they are never
// read. Rung 6 inverts the sense, and that is where the tail word bites.
module Sieve =
    type State = { Words: uint64[]; Size: int }

    let create sieveSize =
        let count = (sieveSize + 1) / 2
        { Words = Array.create ((count + 63) / 64) System.UInt64.MaxValue
          Size = sieveSize }

    let run state =
        let words = state.Words
        let count = (state.Size + 1) / 2
        let limit = int (sqrt (float state.Size))
        let mutable factor = 3

        while factor <= limit do
            let f = factor >>> 1

            if words.[f >>> 6] &&& (1UL <<< f) <> 0UL then
                let mutable q = (factor * factor) >>> 1

                while q < count do
                    words.[q >>> 6] <- words.[q >>> 6] &&& ~~~(1UL <<< q)
                    q <- q + factor

            factor <- factor + 2

        state

    let getResults state =
        let count = (state.Size + 1) / 2
        let primes = ResizeArray<int>()

        if state.Size >= 2 then primes.Add 2

        for i in 1 .. count - 1 do
            if state.Words.[i >>> 6] &&& (1UL <<< i) <> 0UL then primes.Add(2 * i + 1)

        List.ofSeq primes
